### PR TITLE
[Performance] Qwen3.6-27B: Added bucketing fix and TP=8 support

### PR DESCRIPTION
**Pull request by atupe-tt**
### Summary

- ~~Added async scheduling~~ - removed since it is dependent on vLLM PRs that are not yet merged
- Adds decode batch bucketing for Qwen3.6-27B. The model uses the smallest suitable batch size, so a single request stays fast even when the server supports high concurrency.
- Compiles all batch sizes before trace capture and safely keeps separate traces for each size.
- Preserves shared GDN and device state when the batch size changes.
- Fixes unseeded on-device sampling, which could repeat the same random result.
- Adds TP=8 support, including KV-head replication and TP-specific prefill tuning.
- Adds regression tests and CI coverage.
- Async scheduling is deferred until the required vLLM changes are ready.

### Changes to common files

- `models/common/sampling/generator.py`: Adds optional batch-size keys for sampling traces. Models that do not use bucketing keep the existing behavior.
- `models/tt_transformers/tt/generator.py`: Adds optional two-stage trace setup, correct seed initialization, and cleanup for all bucket traces. Buffer reuse is opt-in for Qwen.
- `models/common/warmup/warmup_utils.py`: Lets Qwen compile all batch sizes before trace capture. The default path is unchanged.
- `models/tt_transformers/tt/model_config.py`: Adds opt-in KV-head replication for TP configurations with more devices than KV heads. Other models still use the existing validation.
- `models/demos/utils/device_sku.py`: Adds the P150×8 LoudBox SKU mapping.
- `.github/workflows/vllm-model-tests-impl.yaml`: Disables thinking in the coherence check and increases its output limit.
- Common sampling tests verify that bucketed traces stay isolated without changing other models.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:atupe/qwen36-bucketing-fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:atupe/qwen36-bucketing-fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:atupe/qwen36-bucketing-fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:atupe/qwen36-bucketing-fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:atupe/qwen36-bucketing-fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:atupe/qwen36-bucketing-fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:atupe/qwen36-bucketing-fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:atupe/qwen36-bucketing-fix)

### CODE REVIEW

**Comment by tchedaTT** [comment]
I found a couple issues here. Below is an AI summary of my findings:

Overall, the performance direction makes sense: bucket decode widths, keep sampled tokens and positions resident, and enable steady async decode. I think several correctness and integration requirements need to be addressed before merging.

## Trace warmup and allocation safety

According to the trace-correctness requirements documented in #50156, every program and shape that may allocate from the program cache should be compiled before the first trace is captured.

The current bucket warmup invokes the base warmup separately for each width. Because the base warmup compiles and captures together, this produces approximately:

```text
compile B1 → capture B1
compile B2 → capture B2
compile B4 → capture B4
...
```

It should instead be:

```text
compile B1 → compile B2 → compile B4 → ...
capture B1 → capture B2 → capture B4 → ...
```

Otherwise allocations caused by later compilation can overlap addresses baked into earlier traces.

The new test itself acknowledges this problem by saying not to replay earlier traces after later H2D allocations because doing so can corrupt memory or hang the board: [[test_decode_bucketing.py](https://github.com/tenstorrent/tt-metal/blob/7223bf26bb992260f0eb2b095a34c1666356f706/models/demos/blackhole/qwen36/tests/test_decode_bucketing.py#L639-L671)](https://github.com/tenstorrent/tt-metal/blob/7223bf26bb992260f0eb2b095a34c1666356f706/models/demos/blackhole/qwen36/tests/test_decode_bucketing.py#L639-L671).

Related requirements:

- Keep every trace input/output buffer alive and address-stable.
- Register intentionally overwritten trace inputs/outputs with the trace safety tracker where applicable.
- Maintain separate trace keys for every real control path. The per-bucket sampling-trace namespacing added here is the right idea.
- Release every per-bucket decode and sampling trace during teardown, rather than only whichever store is currently installed on `self.trace_ids_decode`.
- A test showing that four traces fit in the trace region is useful, but it does not prove that earlier traces remain valid after later compilation or allocation.

## Bucketing can be internal, but resident state must survive bucket changes

The power-of-two bucket selection itself can remain internal to tt-metal. vLLM already supplies padded tokens, positions, page tables, active-slot information, and `slot_remap`; it should not need to know the model’s bucket sizes.

However, every bucket currently owns distinct traced token/position/RoPE buffers. Switching buckets therefore changes the buffers that are supposed to contain the authoritative resident state.

One of the following is required on every switch:

1. Drain the pending step and reload authoritative host inputs.
2. Copy/remap authoritative state device-to-device into the incoming bucket.
3. Use canonical resident token/position buffers shared by all bucket traces.

The current `_force_host_decode_tokens` approach is problematic:

- Reloading from the host is safe only after vLLM has drained and made host state authoritative.
- The model makes the bucket decision after vLLM’s overlap/drain decision.
- The flag is only consumed inside the base-generator block guarded by `reset_batch or mode_switched`, and is then cleared unconditionally. A stable bucket switch can therefore ignore it.

A concrete sequence is:

```text
layout change + slot_remap
→ force full-width bucket
→ transition is drained/reloaded
→ next step has no slot_remap
→ return to smaller bucket
→ vLLM sees a stable device-sampling chain
→ reload_inputs=False
→ smaller bucket’s token/position buffers are stale
```

The alternating-width test proves that each sampling trace remains callable and returns in-range token IDs. It does not prove token, position, RoPE, KV, or GDN continuity across the switch.

Please add a test that carries known state through something like `1 → 8 → 1` or `8 → 2 → 8` and compares the actual output/state against an always-reloaded reference.

My preference is to solve this internally with canonical buffers or device-to-device state migration. If a bucket switch requires a host reload, it must be visible to vLLM before the drain/overlap decision.

## This has an actual tenstorrent/vllm#449-style stale-input path

This is not only a request to clean up the code by rebasing onto the new reload contract. The current implementation is exposed to the stale-input ordering problem described in tenstorrent/vllm#449:

```text
previous device-sampling decode is pending
→ host token/position is one step stale
→ vLLM makes its overlap/drain decision
→ _update_states() subsequently discovers a layout change
→ reset_batch=True
→ tt-metal interprets reset_batch as a full input reload
→ stale host token/position can overwrite authoritative device state
```

The shared generator attempts to compensate by reading resident device token/position state, heuristically merging it with host state, applying `slot_remap`, and then reloading. That is itself model-side reload policy and does not establish a general correctness guarantee.

Bucketing introduces another version of the problem: the model decides locally that an incoming bucket needs refreshing, but vLLM did not know that when deciding whether the pending step needed to be drained. Depending on the transition, this can result in either stale host state being reloaded or stale per-bucket device buffers being replayed.

tenstorrent/vllm#444 mitigates some layout-change cases by draining pending work, but it does not establish the complete reload contract. The companion tenstorrent/vllm#454 is also currently based on a commit before tenstorrent/vllm#444 and does not contain it.

This PR should therefore be rebased onto #51646 and tested with tenstorrent/vllm#456. That pair makes vLLM the owner of four independent commands:

```text
reload_inputs
reload_page_table
reload_sampling_params
reset_sampling_state
```

Qwen should execute those commands and should not infer additional host reloads. In particular:

- Page-table-only updates must preserve token/position residency.
- Host/device sampling transitions require a one-time full refresh.
- RNG reset must be independent from forward-input reload.
- `slot_remap` must move sampling state and Qwen GDN recurrent/conv state before decode.
- A no-reload call should assert that device sampling and resident token feedback are supported.
- The overlap/drain decision must use the same transition information that produces the reload commands.

The GDN positional-argument issue also needs fixing: `slot_remap` is accepted positionally by the base API, but the Qwen GDN remap currently reads only `kwargs`.

## Device continuity and async capability

`device_decode_continuity` should not be an independently gated operating mode on the supported Qwen3.6 configuration.

The decode graph can always use device position indices, gather RoPE on device, and advance `current_pos` and RoPE with `plus_one`, regardless of where sampling happens.

The reload plan determines whether those updates survive into the next step:

```text
Device-sampling steady step:
    reload_inputs=False
    sampled token is written into the resident token buffer
    device-advanced position/RoPE remain authoritative
    next decode may overlap

Host-sampling step:
    reload_inputs=True
    host token, position and RoPE inputs overwrite resident values
    the trace may still advance position/RoPE
    the next host reload simply overwrites them again
    no steady decode overlap
```

Device-side position advancement therefore does not need its own environment-variable gate. The same mechanism is safe in both cases; only the source of the next token and the reload plan differ.

For TT’s current contract, useful async decode follows directly from device sampling:

```text
device sampling + no reload/reset
→ resident decode
→ async decode overlap
```

vLLM already decides per step whether device sampling is possible. Structured output, unsupported logits processing, logprobs, bad words, etc. cause host-sampling fallback, and those steps receive `reload_inputs=True`.

Accordingly:

- Remove `TT_QWEN36_DEVICE_DECODE` as a separate continuity/async gate.
- Keep device position/RoPE advancement enabled on the supported Qwen3.6 path.
- Advertise async whenever device sampling is advertised.
- Let vLLM’s explicit reload plan decide whether resident state is preserved.
- Assert that `reload_inputs=False` is only used with device sampling and a supported resident token-feedback path.

If the environment variable remains as a kill switch, it must disable the entire device-sampling/async capability before vLLM accepts the configuration. It cannot leave vLLM advertising async while device continuity is disabled internally.

The sampler-shaped token buffer is a separate implementation detail. It is required when the sampler writes the next token directly into the decode input. If the shared single-device 9B path cannot consume that buffer shape, token-feedback buffer selection should be separated from position/RoPE advancement rather than disabling all device continuity.

For Qwen3.6-27B’s supported P150x4 deployment, both can be enabled unconditionally; the per-step vLLM commands determine whether state is retained or overwritten.

## Why on-device sampling is currently TP-only

Position and RoPE advancement are not inherently TP-only. The restriction comes from the current on-device sampler.

Qwen has a 248,320-token vocabulary, while the sampling Top-K path supports at most 64K logits per shard. On P150x4:

```text
248,320 / 4 = 62,080 logits/device
```

which fits.

The implementation should document and validate that on-device sampling/async requires the certified TP topology. The static capability currently says `supports_sample_on_device=True`, while the instantiated sampler has a stricter topology/vocabulary predicate and can be `None`. This matters because the same wrapper also serves the single-device 9B configuration.

At minimum, requesting device sampling on an unsupported topology should fail clearly during startup rather than later when `self.sampling` is found to be `None`.

## Sampling seed issue

The seed bug fixed here is real, but specifically concerns unseeded (`seed=None`) decode-only sampling:

1. Prefill samples on the host, so device sampling state is not initialized.
2. The first device decode has `reset_batch=True` and `seed=None`.
3. `SeedManager.seeds[slot]` is already `None`.
4. `reset_seed_from_slots_if_needed()` sees no seed change and does nothing.
5. `get_new_values()` incorrectly sees the unseeded RNG as already steady and does not initialize/transition the device RNG.
6. The same random draw can be replayed on every decode.

The regression test is useful and isolates this from sampling-trace correctness by disabling traces.

However, the final fix should follow the tenstorrent/vllm#449 contract rather than tying RNG reset to the overloaded `reset_batch` hint. #51646 contains the stronger form: `reset_sampling_state=True` invokes `reset_decode_batch()`, which freshly initializes unseeded identity slots and clears removed slots independently of forward-input reload.

Please preserve the regression test when rebasing.

This is separate from explicit request-seed determinism. Explicit seeds additionally require their RNG state/counter to follow the request across slot movement and remain aligned with its decode position.

## vLLM companion dependency

tenstorrent/vllm#454 is effectively a companion PR because it supplies the `empty_slots` and `slot_remap` information needed to place and move Qwen GDN state correctly.

That dependency should be stated explicitly and integrated with tenstorrent/vllm#456. `slot_remap` must reach Qwen model state regardless of whether the particular step samples on the host or device; GDN state affects the forward pass independently of sampling location.

Without correct `empty_slots`, a prefill arriving while other requests are decoding can overwrite a live GDN row. Without reliable `slot_remap`, a request’s token/KV/sampling state and its GDN recurrent state can end up in different rows.

## Testing and CI

The new regressions should be registered in tiered CI. Coverage should include:

- All bucket widths precompiled before any trace capture.
- Replay of every earlier bucket after all traces have been captured.
- Actual token/position continuity across bucket switches.
- Bucket switch immediately following `slot_remap`.
- Page-table-only refresh without token/position reload.
- Device→host→device sampling transitions.
- Seeded requests across slot movement.
- Fresh `seed=None` requests reusing identity slots.
- GDN state movement under concurrency/condensation.
- Teardown releasing every bucket’s trace IDs.

The current valid-token alternating-width test and trace-region-fit test do not cover these state-continuity requirements, and the new tests do not appear to be registered in the Qwen tiered-CI entry.

Any permanently skipped board-hanging test should also reference a tracking issue so the failure does not become an unowned permanent skip.

In summary, the performance direction looks reasonable, and the unseeded RNG reproducer is valid. The main blockers are:

1. Precompile all bucket variants before capturing any trace.
2. Preserve authoritative resident state across bucket changes.
3. Adopt the explicit tenstorrent/vllm#449 reload/reset contract.
4. Remove the independent device-continuity/async environment gate.
5. Validate the TP-only on-device sampling topology.
6. Integrate the required vLLM slot-management changes.
7. Add state-continuity and registered CI coverage.

**Review by viktorpusTT — APPROVED** [review]

**Comment by atupe-tt** [comment]
@tchedaTT Thanks for the review! I have addressed most of the above comments. Remaining comments include the ones that depend on the referenced draft PRs in vLLM repo. I will wait till those get merged. 
cc: @ssinghalTT @viktorpusTT 

### Update 08/04/26
@tchedaTT I have removed async_scheduling for now. Will add it later when the corresponding vLLM PRs are merged

**Review by mtairum — CHANGES_REQUESTED** [review]
Blocking this PR due to changes in `.github/workflows/vllm-nightly-tests-impl.yaml`.

We are currently refactoring that pipeline in this PR: https://github.com/tenstorrent/tt-metal/pull/50910

Aim to merge later today or tomorrow. Once it gets on main, please update the vllm changes and ping me on slack to approve 👍

**Comment by atupe-tt** [comment]
Remaining integration depends on:

- [tenstorrent/vLLM #458](https://github.com/tenstorrent/vllm/pull/458) - explicit reload planning; supersedes https://github.com/tenstorrent/vllm/pull/456 and includes https://github.com/tenstorrent/vllm/pull/444.
- [tenstorrent/vLLM #454](https://github.com/tenstorrent/vllm/pull/454) - required `empty_slots`/`slot_remap` support; needs integration with https://github.com/tenstorrent/vllm/pull/458.
- [tt-metal #50910](https://github.com/tenstorrent/tt-metal/pull/50910) - vLLM CI pipeline migration.
- [tt-metal #50156](https://github.com/tenstorrent/tt-metal/pull/50156) - trace-allocation safety tracker for final validation.

After these land, I will rebase this PR, update the CI registration, and run the paired vLLM/tt-metal regressions.

cc: @mbahnasTT @ssinghalTT

**Comment by mtairum** [comment]
@atupe-tt the vLLM PR is in merge queue: https://github.com/tenstorrent/tt-metal/pull/50910

The queue is slow today, so it should be merged in the next few hours.

**Comment by atupe-tt** [comment]
@mtairum Sounds good! Thanks!

**Comment by atupe-tt** [comment]
/codeowners ping
Could you please review the remaining changes, excluding the vLLM nightly change?

**Review by mtairum — APPROVED** [review]

File: models/demos/utils/device_sku.py [hunk 1]
```diff
@@ -25,6 +25,7 @@ def cluster_type_to_sku_name(cluster_type: ttnn.cluster.ClusterType) -> str:
         ttnn.cluster.ClusterType.P150: "bh_p150",
         ttnn.cluster.ClusterType.P300: "bh_p300",
         ttnn.cluster.ClusterType.P300_X2: "bh_quietbox_2",
+        ttnn.cluster.ClusterType.P150_X8: "bh_loudbox",
```
Comment (mtairum) [inline models/demos/utils/device_sku.py:28]: Any reason why we are adding this SKU here?

It's not an official SKU anymore.

Comment (atupe-tt) [reply models/demos/utils/device_sku.py:28]: I added it to support the TP=8 implementation for this model

**Review by uaydonat — APPROVED** [review]

**Review by djordje-tt — APPROVED** [review]

**Review by ssinghalTT — APPROVED** [review]
